### ISSUE TITLE
Paniced on a Sanity Check - Expected storeHeight and stateHeight to match

### ISSUE THREAD

**Issue by baabeetaa** [tendermint/tendermint#388]
How to fix this?!

```
NOTE[02-01|16:24:00] ABCI Handshake                           module=state appHeight=210486 appHash=0A2973AF199B84C07933D58A5A89422228FF4399
NOTE[02-01|16:24:00] ABCI Replay Blocks                       module=state appHeight=210486 storeHeight=210488 stateHeight=210487
panic: Paniced on a Sanity Check: Expected storeHeight (210488) and stateHeight (210487) to match.
```

**Comment by ethanfrey** [comment]
Can you please provide more details?
Which version of tendermint? 0.8.0? Dev?
Which abci app? From tendermint or self-writen?
Is it using merkleeyes to persist data? If so, which commit?

I ask as there were quite a few changes recently. It seems that the abci app didn't persist it's state properly, but to debug better one needs more info.

It does seem to be a bug. But to locate the issue, we need more info.

Thanks

**Comment by baabeetaa** [comment]
Hi Ethan, got this bug after restarting services ( app & TM )
- TM 0.8
- self-written app
- merkleeyes

**Comment by ethanfrey** [comment]
Okay, thanks for the info.  Maybe @ebuchman has come across this before?  I have only seen a similar error when I change the data between app starts.  (Like I change the genesis block and such), then I need to do `tendermint unsafe_reset_all`

If you don't mind sending the `merkleeyes.db` (zipped up), I could take a deeper look.  This is the level db with all the state, which apparently is returning the bad state.  I have written a few apps (mostly based on basecoin, and thus merkleeyes) and never had this issue on restart.  Like I said, I got this when actually changing state on the disk between runs, in which case some failing sanity check is proper.

If you don't feel comfortable sharing that info, I understand, will await insight from ebuchman and then see if I can reproduce.

**Comment by baabeetaa** [comment]
I uploaded the data here: https://drive.google.com/file/d/0B7rMi8tL1VG9WlRCYkJCdlltS3c/view?usp=sharing

There is the VPS box running, let me know if you need ssh to.
This is the testbox, so I can reset all the data. However, I like to known if It could be fixed. And what is the best way to shutdown app & TM to avoid this issue.

**Comment by ebuchman** [comment]
I believe this happens if you shut down the app first. If you shut down tendermint first, this should not happen.  Either way, we need to fix this :)

**Comment by baabeetaa** [comment]
Thanks @ebuchman , my bad habit is stop the app first then TM will be stopped automatically after that.

**Comment by ebuchman** [comment]
Not a problem - we need to support that. Allowing the app to stop and restart is a big open issue right now (I tried to solve it before, but wasn't a good approach, and we put it on the backburner). https://github.com/tendermint/tendermint/issues/243

**Comment by faddat** [comment]
@baabeetaa's talking of course about github.com/baabeetaa/glogchain, if you want to have a gander at the code.

**Comment by ebuchman** [comment]
Should be fixed with v0.9.0 on master. please reopen if it's not
